drivers/fpga/imx_m4.c

```diff
@@ -360,6 +360,12 @@ static int m4_boot(struct spi_flash *flash, size_t size, struct hash *p)
 	memset((void *) M4_BASE, 0x00, M4_SIZE);
 	spi_flash_erase(flash, 0, M4_SIZE);
 
+	/* invalidate the RPMB information so we trigger an upgrade on the
+	   next boot avoiding an infinite loop */
+	memset(hash.value, 0x00, sizeof(hash.value));
+	hash.len = 0;
+	update_secure_hash(&hash, 0);
```

Comment:
If I read this correctly any error will also erase the hash stored in RPMB, which can be a way to allow booting older images (rollback attack), but I guess this is fine as the way to protect against a rollback would be by storing the timestamp that gets provided by the fitimage itself.

Reply:
Wouldn't it be better to just assume that if the SPI is 0 (after erase) and hash in fit is same as hash in RPMB then we perform another "update"?

Reply:
so, any error detected during the boot sequence  - right after we have flashed a new image OR we have decided there is no need to upgrade and we are going for a boot -we erase the flash/rpmb to accept a new image.  With this we can detect errors done during the flash write sequence or image corruptions

sorry not sure what you mean by perform another update? you mean only accepting firmwares with the secure hash?

Reply:
I'm just thinking if there is a need to also erase the rpmb hash as we could simply force another flash write if hash is different or if the content of the flash is 0 (which will be true after a bad boot try, as you already had the logic in place for erasing it).

Reply:
we could retry the operation a number of times but how do we terminate so we dont get into an infinite loop? at some point I think we need to clean the hash

Reply:
I guess this is fine, we can protect against rollback separately later (which will cover the entire fitimage). @mike-scott any comments here?

Reply:
anything?

Reply:
If we're erasing SPI flash here, we probably need to reset the hash value as well.  Makes sense.